app/javascript/app/document-capture/models/data-url-file.js

```diff
@@ -0,0 +1,29 @@
+/**
```

Comment:
There were a few alternatives I had considered before arriving at this approach:

1. Represent the value as a string, and assign `name` as a property of the string. The language supports this, though it's very much a hack to assign a non-standard property to a string this way. 
   - Example: `Object.assign( 'data:image\/png;base64,', { name: 'demo.png' } ) // String {"data:image/png;base64,", name: "demo.png"}` 
2. Allow a step to define its serialization behavior, similar to what was done with validation for `step.isValid`
   - Example: `step.serialize`
3. Don't bother to maintain the `name` value. This would require altering the design, since it is presented to the user in the preview (see https://github.com/18F/identity-idp/pull/3999#issuecomment-668097755)
4. Create the object inline with a `toJSON` property
   - Example: `JSON.stringify( { data: 'data:image\/png;base64,', name: 'demo.png', toJSON: () => 'data:image\/png;base64,' } ) // '"data:image/png;base64,"'`

Reply:
What if we put the name inside the URL?

I just messed around in Ruby and I think we can use the URL fragment to keep the name if we want to:

```ruby
uri = URI('data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==')
uri.opaque
# => text/plain;base64,SGVsbG8sIFdvcmxkIQ==


uri_with_fragment = URI('data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==#name=xyz')
uri_with_fragment.opaque
# => text/plain;base64,SGVsbG8sIFdvcmxkIQ==
uri_with_fragment.fragment
# => 'name=xyz'
```

Reply:
Ah, this reminds me that there was another alternative I'd explored, which is to try to find some way to embed this as metadata within the URL itself, in the same line of thinking as your suggestion. Though I was looking at the ["parameter" fragment](https://tools.ietf.org/html/rfc2397#section-3) of the Data URL syntax. It wasn't clear to me that it could be used this way.

A hash fragment may work as well. It seems to be similarly agreeable to JavaScript's `URL` constructor:

![url with hash](https://user-images.githubusercontent.com/1779930/89441067-96f54780-d71a-11ea-9e29-e6af1bfa8b79.png)

And as far as I understand, it should be considered a valid URL.

There's still a few open questions:

- Even if it's valid, will the Rails application (or, more importantly, the third-party vendor) tolerate it?
- Since this data is only relevant for the client, is it less-than-ideal that we're also sending it as part of the payload to the server in the submission?
- We'd have to implement something to manage parsing the pieces from the hash fragment (some prior art for this in the `useHistoryParam` hook)
- Is it likely we might need to have a pattern to support the option to define custom serialization behavior for any other form values?

I don't feel strongly one way or the other, but I'm not overly troubled by the original implementation proposed here. It seems to me as a valid use of the `toJSON` behavior for data we're expecting to be sent as a JSON payload.

Reply:
OK! I thought we wanted to try to preserve the filename but if it's not needed at all, let's go with the `toJSON` then

Reply:
Yet another option here is that we can maintain the value as an object in the browser, then adapt the server API endpoint to receive the full object and use only the `data` property of that object. In these scenarios, I tend to try to not let the client-side requirements drive the server-side API design. It's an option to consider down the line in case this causes headaches for us in the future.